Vectorize Enumerable.Range initialization, take 2

neon-sunset opened the pull request:
Previous attempt: https://github.com/dotnet/runtime/pull/80633

This one fixes the regression on x86_64 by avoiding an `SDIV` and having smaller method body.

Diffs:
- x86_64 https://www.diffchecker.com/uz0rCGdv/
- Aarch64 https://www.diffchecker.com/2Czwh3OI/

gfoidl commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,52 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
+
+                int stride = Vector256<int>.Count;
+
+                Vector256<int> init = Vector256.Create(
+                    (ReadOnlySpan<int>)new int[] { 0, 1, 2, 3, 4, 5, 6, 7 });
```
```suggestion
                Vector256<int> init = Vector256.Create(0, 1, 2, 3, 4, 5, 6, 7);
```
The a [constant](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbho1iAZiakGAYQYBvGg11NlANRhgM0UgFYkAHnEYAfAwCy5ABQBKLTr3ejJs5ZseB3FeDAYAXgZfUygLJBY1WGwMGBcXNhhsABMAeS4AGwBPAGUAB2wuQPs3LhgAdwZbAG0AXS0GOjQGci7SLsUu9AZzLqQugHYGAF83OWpvb2JJkIw57ymvXU39KOMYuKqHR1J3T3mF3Wj/a1tgrlCI3b9YywSklJdO7t7+wa6RhhjBjjWbbRbLe6rbYbahTIA===) will be emitted and used.

neon-sunset replied:
Nice, it seems it wasn't the case in the past but now the codegen is optimal in either case (in Tier 1). I haven't looked at Native AOT though. Fixed.

gfoidl commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,52 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
+
+                do
+                {
+                    value.StoreUnsafe(ref MemoryMarshal.GetReference(destination));
```
Get the `ref` once before the loop and use that `ref` throughout.

neon-sunset replied:
Fixed. The codegen did not change though.

EgorBo commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
 
-            private static void Fill(Span<int> destination, int value)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static void Fill(Span<int> destination, int start)
```
why is this better than just Span.Fill ?

neon-sunset replied:
This one is for filling in sequences of e.g. 0, 1, 2, 3, 4, ...N.
Or if you mean why the inline was added - previously `Fill` was getting inlined in Tier 1 and the attribute gives extra help to ensure short ranges do not regress. This is okay here because the caller itself won't get inlined so it only affects the bodies of `ToArray()` and `ToList()`.

EgorBo replied:
Ah I see, I confused it with "fill with value"

EgorBo commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
+                    goto Scalar;
+                }
+
+                int stride = Vector256<int>.Count;
```
Any reason to use V256 instead of V128 to support both x86 and arm?

EgorBo replied:
Ok what's the codegen for e.g. SSE2 (NativeAOT's default target)?

EgorBo replied:
Also, I don't see any guards for SIMD at all - arm32, linux-x86, mono-interp, HWINTRINSIC=0

neon-sunset replied:
> Ok what's the codegen for e.g. SSE2 (NativeAOT's default target)?
Let me check. Also, is it really as low as SSE2? That would be unfortunate.

EgorBo replied:
> > Ok what's the codegen for e.g. SSE2 (NativeAOT's default target)?
> 
> Let me check. Also, is it really as low as SSE2? That would be unfortunate.

Yes, with SSE4.1 as opportunistic check as far as I remember (runtime-check)

neon-sunset replied:
`DOTNET_EnableSSE41=0`
```asm
; 2 inlinees with PGO data; 29 single block inlinees; 11 inlinees without PGO data

G_M000_IG01:                ;; offset=0000H
       push     rsi
       sub      rsp, 64

G_M000_IG02:                ;; offset=0005H
       mov      esi, dword ptr [rcx+08H]
       mov      edx, dword ptr [rcx+0CH]
       sub      edx, esi
       movsxd   rdx, edx
       mov      rcx, 0x7FFD85693060
       call     CORINFO_HELP_NEWARR_1_VC
       lea      rdx, bword ptr [rax+10H]
       mov      ecx, dword ptr [rax+08H]
       cmp      ecx, 8
       jl       SHORT G_M000_IG06

G_M000_IG03:                ;; offset=002BH
       movd     xmm0, esi
       pshufd   xmm0, xmm0, 0
       movups   xmm1, xmmword ptr [reloc @RWD00]
       paddd    xmm1, xmm0
       movups   xmm2, xmmword ptr [reloc @RWD16]
       paddd    xmm0, xmm2
       movups   xmm2, xmmword ptr [reloc @RWD32]
       movups   xmm3, xmmword ptr [reloc @RWD32]
       align    [8 bytes for IG04]

G_M000_IG04:                ;; offset=0060H
       movups   xmmword ptr [rdx], xmm1
       movups   xmmword ptr [rdx+10H], xmm0
       paddd    xmm1, xmm3
       paddd    xmm0, xmm2
       add      rdx, 32
       sub      ecx, 8
       cmp      ecx, 8
       jge      SHORT G_M000_IG04

G_M000_IG05:                ;; offset=007BH
       movaps   xmmword ptr [rsp+20H], xmm1
       movaps   xmmword ptr [rsp+30H], xmm0
       mov      esi, dword ptr [rsp+20H]

G_M000_IG06:                ;; offset=0089H
       xor      r8d, r8d
       test     ecx, ecx
       jle      SHORT G_M000_IG08
       align    [0 bytes for IG07]

G_M000_IG07:                ;; offset=0090H
       lea      r9d, [rsi+01H]
       mov      r10d, r8d
       mov      dword ptr [rdx+4*r10], esi
       inc      r8d
       cmp      r8d, ecx
       mov      esi, r9d
       jl       SHORT G_M000_IG07

G_M000_IG08:                ;; offset=00A6H
       add      rsp, 64
       pop      rsi
       ret

RWD00   dq      0000000100000000h, 0000000300000002h
RWD16   dq      0000000500000004h, 0000000700000006h
RWD32   dq      0000000800000008h, 0000000800000008h

; Total bytes of code 172
```
Very similar to ARM64. I have added check for V128 acceleration too.

EgorBo replied:
Do we use this approach anywhere else in BCL? Last time we discussed this, V256 via V128x2 was mainly used for testing only

neon-sunset replied:
How do you want the implementation to look like? Full span.IndexOf style with V256 -> V128 -> Scalar blocks?

EgorBo replied:
> How do you want the implementation to look like?

I am fine with the codegen you posted - I am just not aware if this approach is recommended/used anywhere. cc @tannergooding

stephentoub replied:
> How do you want the implementation to look like?

Has something changed since the previous PR?  My feedback there was that without motivating scenarios, this wasn't worth vectorizing at all.  Can you point to places where `Enumerable.Range(...).ToArray/ToList` is being used in production code on perf-sensitive code paths?

neon-sunset replied:
Thank you for reviewing the change. Here are the concerns listed in the original PR that were stating the impl. was not worth it:
- Negative impact on code size
- Negative impact on maintainability
- Negative impact on un-vectorized path

I tried to address the first two by keeping the implementation as simple as possible and evaluating codegen size diff, which is +60B on x86_64 and +100B on Aarch64. The third one is addressed by getting rid of `length % Vector<int>.Count` which was causing regressions on x86_64.

It is true that the pattern most often used in tests, but I don't agree with the assertion that it is not a useful change. There are also references to non-test code that could benefit:
- https://github.com/fabsenet/adrilight/blob/main/adrilight/ViewModel/SettingsViewModel.cs#L55-L56
- https://github.com/SciSharp/Pandas.NET/blob/master/src/Pandas.NET/Methods/Pandas.read_csv.cs#L43
- https://github.com/teenageengineering/videolab/blob/master/Assets/Skinner/SkinnerModel.cs#L67

Matches: https://grep.app/search?q=Enumerable%5C.Range%5C%28%5B%5E%29%5D%2B%5C%29%5C.%28ToArray%7CToList%29%5C%28%5C%29&regexp=true&case=true&filter[lang][0]=C%23 (22 pages of results)

stephentoub commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
             }
 
-            private static void Fill(Span<int> destination, int value)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
I don't think this should be aggressively inlined.

stephentoub commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
-                Fill(array, _start);
+                int start = _start;
+                int[] array = new int[_end - start];
+                Fill(array, start);
```
Does reading _start once instead of twice make a meaningful difference?  Same question below for _end and _start.

stephentoub replied:
Do you have evidence that it actually helps?  From looking at the codegen, this actually looks worse.  Please revert.

stephentoub commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -16,23 +19,55 @@ public override IEnumerable<TResult> Select<TResult>(Func<int, TResult> selector
             {
-                for (int i = 0; i < destination.Length; i++, value++)
+                // The improvements in the compiler now allow us to write the loop as Vector256.
+                // Manually unrolling to Vector128x2 for SSE2/NEON is no longer profitable.
```
Can we just use `Vector<T>`? e.g.
```C#
static void Fill(Span<int> destination, int value)
{
    Debug.Assert(Vector<int>.Count * sizeof(int) * 8 <= 512);

    ref int c = ref MemoryMarshal.GetReference(destination);
    ref int end = ref Unsafe.Add(ref c, destination.Length);

    if (Vector.IsHardwareAccelerated && destination.Length >= Vector<int>.Count)
    {
        var init = new Vector<int>((ReadOnlySpan<int>)new int[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 });
        var increment = new Vector<int>(Vector<int>.Count);
        var current = new Vector<int>(value) + init;

        ref int oneVectorFromEnd = ref Unsafe.Subtract(ref end, Vector<int>.Count);
        do
        {
            current.StoreUnsafe(ref c);
            current += increment;
            c = ref Unsafe.Add(ref c, Vector<int>.Count);
        }
        while (!Unsafe.IsAddressLessThan(ref oneVectorFromEnd, ref c));

        value = current[0];
    }

    while (Unsafe.IsAddressLessThan(ref c, ref end))
    {
        c = value++;
        c = ref Unsafe.Add(ref c, 1);
    }
}
```

neon-sunset replied:
It is definitely an option, but I initially opted out of it to have better performance on ARM64. As of now, there are pros and cons.

`Vector<T>` variant
Pros:
- Takes advantage of AVX512
- Does not rely on JIT/ILC being able to perfectly optimize V256 as V128x2 for AdvSimd/SSE2
- Consistent with how other vectorized code is written
- Better performance on Mono? (as @EgorBo mentioned, Mono regressions are an issue that needs to be tracked)

Cons:
- Does not help 0..15 i.e. `Enumerable.Range(0, 10).ToArray()` , which is fairly common, on AVX512
- Has reduced gains on wide ARM64 cores (M series can dispatch V128x2 loop iteration in a single cycle, ARM Cortex-X3/4 should be able to do the same). This also applies to default NativeAOT target which would use 128x2 with SSE2. I expect x86_64 behavior to be very similar.

Do you think `Vector<T>` is better for the time being?

stephentoub replied:
> Takes advantage of AVX512

Not by default.  By default, `Vector<T>` will be either 128-bits or 256-bits.

> Do you think Vector<T> is better for the time being?

Yes

neon-sunset committed:
Vectorize Enumerable.Range initialization

neon-sunset commented:
@stephentoub I have addressed the feedback but kept `AggressiveInlining` because `Fill` will get inlined into `ToArray`, `ToList` and `CopyTo` which are behind virtual calls and won't get inlined. Diff example: https://www.diffchecker.com/MWZRN2Sy/

stephentoub commented:
> which are behind virtual calls and won't get inlined further

They can be with dynamic pgo. Please remove the aggressive inlining.

neon-sunset commented:
> They can be with dynamic pgo. Please remove the aggressive inlining.

Sorry for the confusion, I missed the attribute when force-pushing. The change is already without it, as you requested.

stephentoub commented on src/libraries/System.Linq/src/System/Linq/Range.SpeedOpt.cs:
```diff
@@ -33,9 +39,33 @@ public List<int> ToList()
+
+                if (Vector.IsHardwareAccelerated && destination.Length >= Vector<int>.Count)
+                {
+                    Vector<int> init = new((ReadOnlySpan<int>)new int[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 });
+                    Vector<int> current = new Vector<int>(value) + init;
+                    Vector<int> increment = new(Vector<int>.Count);
```
I know why target-typed new isn't being used for the second line, but it looks inconsistent to use it for the first and third but not the second.  Please make them all `new Vector<int>`.

stephentoub committed:
Merge branch 'main' into simd-range-init

stephentoub committed:
Address PR feedback

stephentoub commented:
Thanks.

neon-sunset commented:
@stephentoub thank you for merging this and apologies for not addressing the feedback previously!
The merged change yields 1.5-4x speed-up vs 7.0 depending on the range length (and up to 10x for List thanks to CollectionsMarshal.SetCount):
### x86_64
```
BenchmarkDotNet v0.13.6, Windows 11 (10.0.22621.1992/22H2/2022Update/SunValley2)
AMD Ryzen 7 5800X, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.100-rc.1.23371.8
  [Host]        : .NET 8.0.0 (8.0.23.37103), X64 RyuJIT AVX2
  .NET 7.0      : .NET 7.0.9 (7.0.923.32018), X64 RyuJIT AVX2
  .NET 8.0      : .NET 8.0.0 (8.0.23.37103), X64 RyuJIT AVX2
  NativeAOT 8.0 : .NET 8.0.0-rc.1.23371.3, X64 NativeAOT AVX2
```

| Method |           Job |       Runtime | Length |        Mean |     Error |    StdDev | Ratio |
|------- |-------------- |-------------- |------- |------------:|----------:|----------:|------:|
|  Array |      .NET 7.0 |      .NET 7.0 |     10 |    14.09 ns |  0.125 ns |  0.111 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |     10 |    11.89 ns |  0.124 ns |  0.104 ns |  0.84 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |     10 |    11.66 ns |  0.142 ns |  0.133 ns |  0.83 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |     10 |    22.10 ns |  0.196 ns |  0.183 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |     10 |    15.41 ns |  0.161 ns |  0.151 ns |  0.70 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |     10 |    20.20 ns |  0.132 ns |  0.103 ns |  0.92 |
|        |               |               |        |             |           |           |       |
|  Array |      .NET 7.0 |      .NET 7.0 |    100 |    45.49 ns |  0.481 ns |  0.376 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |    100 |    22.23 ns |  0.264 ns |  0.234 ns |  0.49 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |    100 |    22.36 ns |  0.374 ns |  0.331 ns |  0.49 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |    100 |   112.15 ns |  1.528 ns |  1.276 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |    100 |    25.09 ns |  0.215 ns |  0.191 ns |  0.22 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |    100 |    32.31 ns |  0.378 ns |  0.335 ns |  0.29 |
|        |               |               |        |             |           |           |       |
|  Array |      .NET 7.0 |      .NET 7.0 |   1000 |   356.33 ns |  3.589 ns |  3.358 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |   1000 |   117.84 ns |  2.250 ns |  2.105 ns |  0.33 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |   1000 |   121.89 ns |  1.932 ns |  1.807 ns |  0.34 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |   1000 |   984.15 ns |  7.978 ns |  6.662 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |   1000 |   125.00 ns |  2.064 ns |  1.931 ns |  0.13 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |   1000 |   133.36 ns |  1.298 ns |  1.214 ns |  0.14 |
|        |               |               |        |             |           |           |       |
|  Array |      .NET 7.0 |      .NET 7.0 |  10000 | 3,209.78 ns | 21.071 ns | 19.709 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |  10000 |   958.51 ns | 18.953 ns | 21.067 ns |  0.30 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |  10000 |   947.25 ns | 16.773 ns | 14.007 ns |  0.29 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |  10000 | 9,387.75 ns | 86.637 ns | 81.040 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |  10000 |   965.64 ns | 11.597 ns | 10.848 ns |  0.10 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |  10000 |   970.69 ns | 17.986 ns | 16.825 ns |  0.10 |

### Arm64
```
BenchmarkDotNet v0.13.6, macOS Sonoma 14.0 (23A5286i) [Darwin 23.0.0]
Apple M1 Pro, 1 CPU, 8 logical and 8 physical cores
.NET SDK 8.0.100-rc.1.23371.8
  [Host]        : .NET 8.0.0 (8.0.23.37103), Arm64 RyuJIT AdvSIMD
  .NET 7.0      : .NET 7.0.1 (7.0.122.56804), Arm64 RyuJIT AdvSIMD
  .NET 8.0      : .NET 8.0.0 (8.0.23.37103), Arm64 RyuJIT AdvSIMD
  NativeAOT 8.0 : .NET 8.0.0-rc.1.23371.3, Arm64 NativeAOT AdvSIMD
```

| Method |           Job |       Runtime | Length |        Mean |     Error |    StdDev | Ratio |
|------- |-------------- |-------------- |------- |------------:|----------:|----------:|------:|
|  Array |      .NET 7.0 |      .NET 7.0 |     10 |    39.13 ns |  0.506 ns |  0.395 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |     10 |    13.30 ns |  0.018 ns |  0.016 ns |  0.34 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |     10 |    16.24 ns |  0.028 ns |  0.023 ns |  0.41 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |     10 |    26.45 ns |  0.037 ns |  0.035 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |     10 |    17.99 ns |  0.031 ns |  0.029 ns |  0.68 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |     10 |    22.34 ns |  0.047 ns |  0.044 ns |  0.84 |
|        |               |               |        |             |           |           |       |
|  Array |      .NET 7.0 |      .NET 7.0 |    100 |    58.72 ns |  0.163 ns |  0.152 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |    100 |    31.41 ns |  0.058 ns |  0.048 ns |  0.53 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |    100 |    32.82 ns |  0.082 ns |  0.073 ns |  0.56 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |    100 |   109.95 ns |  0.158 ns |  0.140 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |    100 |    37.40 ns |  0.101 ns |  0.094 ns |  0.34 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |    100 |    40.23 ns |  0.375 ns |  0.351 ns |  0.37 |
|        |               |               |        |             |           |           |       |
|  Array |      .NET 7.0 |      .NET 7.0 |   1000 |   469.01 ns |  3.150 ns |  2.947 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |   1000 |   266.25 ns |  0.453 ns |  0.424 ns |  0.57 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |   1000 |   249.36 ns |  0.383 ns |  0.319 ns |  0.53 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |   1000 |   856.08 ns |  3.760 ns |  3.140 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |   1000 |   278.62 ns |  0.551 ns |  0.460 ns |  0.33 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |   1000 |   259.52 ns |  0.809 ns |  0.756 ns |  0.30 |
|        |               |               |        |             |           |           |       |
|  Array |      .NET 7.0 |      .NET 7.0 |  10000 | 4,111.95 ns | 68.902 ns | 64.451 ns |  1.00 |
|  Array |      .NET 8.0 |      .NET 8.0 |  10000 | 2,420.28 ns | 13.809 ns | 12.917 ns |  0.59 |
|  Array | NativeAOT 8.0 | NativeAOT 8.0 |  10000 | 2,284.52 ns |  7.150 ns |  6.338 ns |  0.56 |
|        |               |               |        |             |           |           |       |
|   List |      .NET 7.0 |      .NET 7.0 |  10000 | 7,661.80 ns | 19.279 ns | 18.034 ns |  1.00 |
|   List |      .NET 8.0 |      .NET 8.0 |  10000 | 2,581.48 ns | 15.425 ns | 14.429 ns |  0.34 |
|   List | NativeAOT 8.0 | NativeAOT 8.0 |  10000 | 2,487.72 ns | 11.772 ns | 10.436 ns |  0.32 |
